rust/common-graph/src/graph.rs

```diff
@@ -0,0 +1,734 @@
+use super::PortDetails;
+use super::{CommonGraphError, ConnectionDetails, Result};
+use crate::storage::GraphData;
+use crate::utils::{is_empty, non_unique_entries};
+use crate::{GraphProcessor, GraphProcessorItem, Processor};
+use std::collections::HashSet;
+use std::fmt::Debug;
+
+/// A reference to a port via the node index and port name.
+pub type PortRef = (usize, String);
+/// An edge in a graph containing the source `0` and target `1` ports.
+pub type Edge = (PortRef, PortRef);
+
+/// Representing a distinction between input and output ports.
+#[derive(strum::Display, Debug, PartialEq)]
+pub enum PortType {
+    /// An input port.
+    #[strum(to_string = "input")]
+    Input,
+    /// An output port.
+    #[strum(to_string = "output")]
+    Output,
+}
+
+/// The internal node of a [`Graph`].
+#[derive(Debug)]
+pub struct Node<T> {
```

Comment:
> We should get an issue filed to track the eventual task of describing how an already-built graph may be (de)serialized.

If a given graph of `Graph<T: Serialize + Deserialize>`, this should be rather transparent from the inputs of `Graph::new` which are "raw" index-based inputs

> 
> Relatedly: given an immutable graph, I want to be able to re-wrap it in a builder and fork it.

A few things:
* `Graph` defines a graph by nodes and connections by node index/port name. `GraphBuilder` defines connections by node *name* for ease of building, and gets converted to index when building the `Graph`.
* We could get rid of node "names" entirely in the underlying `Graph`, deferring to e.g. `T: Display` when needed in errors
* Not sure if we'd even want `GraphBuilder` outside of tests if the runtime primarily constructs graphs via JSON description
* When would a graph be forked into a builder? Given the additive stateful nature of `GraphBuilder` there are only a few mutations that would even work (fan-out to a new node into an empty graph output node)

Reply:
The user story for remix is something like:

> As a Fabric user, when I work with a recipe that isn't exactly what I need, I want to be able to make small tweaks to the recipe so that I can make it perfect for my use case

Basically I want easy way to get back into builder mode (it doesn't necessarily need to use the same builder) for the purpose of modifying the graph.

Reply:
Anyway, the request is that we file some issues for further discussion